File: pipeline/poles/classes.py
"""Distance class table shared by the explore tiles, the detail rasters and the site (spec 3.4).

Class c covers [edges[c], edges[c+1]) metres; the last real class, 253, is open-ended. Two reserved values
sit above the real classes: EDGE for cells whose distance is only a lower bound because the data ends within
edge_mask_m of them, NODATA for water and everything outside the region's data. site/js/classes.js mirrors
this file; tests/test_classes.py compares the two."""
from __future__ import annotations

import math

import numpy as np

EDGE = 254
NODATA = 255
N_CLASSES = 254  # real classes 0..253


def default_edges() -> list[int]:
    edges = list(range(0, 2_500, 50))                # 50 m steps to 2.5 km: classes 0..49
    edges += list(range(2_500, 10_000, 100))         # 100 m steps to 10 km: classes 50..124
    edges += list(range(10_000, 30_000, 250))        # 250 m steps to 30 km: classes 125..204
    edges += list(range(30_000, 60_000, 1_000))      # 1 km steps to 60 km: classes 205..234
    edges += list(range(60_000, 240_000, 10_000))    # 10 km steps to 240 km: classes 235..252
    edges.append(240_000)                            # class 253: 240 km and beyond
    return edges


class ClassTable:
    def __init__(self, edges: list[int] | None = None):
        e = [int(v) for v in (edges if edges is not None else default_edges())]
        if len(e) != N_CLASSES:
            raise ValueError(f"class table needs {N_CLASSES} lower edges, got {len(e)}")
        if e[0] != 0 or any(b <= a for a, b in zip(e, e[1:])):
            raise ValueError("class edges must start at 0 and increase strictly")
        self.edges = e
        self._arr = np.asarray(e, dtype=np.float64)

    def to_class(self, dist_m) -> np.ndarray:
        """Class of each distance in metres, as uint8.

        Feed a block or a window, not a whole continental grid: this allocates arrays the size of its input,
        and the publish stage windows its rasters for exactly that reason."""
        d = np.asarray(dist_m)
        if d.dtype.kind != "f":  # already floating point stays as it is; ints and bools need a float view
            d = d.astype(np.float64)
        if not np.all(np.isfinite(d)) or np.any(d < 0):
            raise ValueError("distances must be finite and non-negative")
        return (np.searchsorted(self._arr, d, side="right") - 1).astype(np.uint8)

    def _check(self, c: int) -> int:
        if not 0 <= c < len(self.edges):
            raise ValueError(f"class {c} is outside 0..{len(self.edges) - 1}")
        return c

    def lower(self, c: int) -> int:
        return self.edges[self._check(c)]

    def upper(self, c: int) -> float:
        c = self._check(c)
        return float(self.edges[c + 1]) if c + 1 < len(self.edges) else math.inf

    def mid(self, c: int) -> float:
        hi = self.upper(c)
        return (self.lower(c) + hi) / 2 if hi != math.inf else float(self.lower(c))

File: site/js/classes.js
// Distance class table, the mirror of pipeline/poles/classes.py (spec 3.4). Keep both in step: the
// pipeline test suite compares them whenever node is on PATH. The published regions.json carries its
// region's edges, so the site builds its table with makeClassTable(edges) rather than hard-coding one.
export const EDGE = 254;
export const NODATA = 255;

const N_CLASSES = 254; // real classes 0..253; EDGE and NODATA sit above them

function range(start, stop, step) {
  const out = [];
  for (let v = start; v < stop; v += step) out.push(v);
  return out;
}

export const CLASS_EDGES = [
  ...range(0, 2500, 50),
  ...range(2500, 10000, 100),
  ...range(10000, 30000, 250),
  ...range(30000, 60000, 1000),
  ...range(60000, 240000, 10000),
  240000,
];

// A table over one region's lower edges. Class c covers [edges[c], edges[c + 1]) metres and class 253 is
// open-ended. Validated exactly as ClassTable does in Python, but throwing RangeError.
export function makeClassTable(edges = CLASS_EDGES) {
  const e = Array.from(edges, Number);
  if (e.length !== N_CLASSES) {
    throw new RangeError(`class table needs ${N_CLASSES} lower edges, got ${e.length}`);
  }
  if (e.some((v) => !Number.isInteger(v))) {
    throw new RangeError('class edges must be whole metres');
  }
  if (e[0] !== 0 || e.some((v, i) => i > 0 && !(v > e[i - 1]))) {
    throw new RangeError('class edges must start at 0 and increase strictly');
  }
  const checkClass = (c) => {
    if (!Number.isInteger(c) || c < 0 || c >= e.length) {
      throw new RangeError(`class ${c} is outside 0..${e.length - 1}`);
    }
    return c;
  };
  // Class of a distance in metres: the last edge not above it.
  const toClass = (distM) => {
    if (!Number.isFinite(distM) || distM < 0) {
      throw new RangeError('distance must be a finite non-negative number');
    }
    let lo = 0;
    let hi = e.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (e[mid] <= distM) lo = mid + 1; else hi = mid;
    }
    return lo - 1;
  };
  const lower = (c) => e[checkClass(c)];
  const upper = (c) => (checkClass(c) + 1 < e.length ? e[c + 1] : Infinity);
  return { edges: e, toClass, lower, upper };
}

// Shortcuts over the default table, so the binary search lives in one place.
const DEFAULT_TABLE = makeClassTable(CLASS_EDGES);

export const toClass = DEFAULT_TABLE.toClass;
export const classLower = DEFAULT_TABLE.lower;
export const classUpper = DEFAULT_TABLE.upper;
